service/r2/routes.go

```diff
@@ -271,3 +275,117 @@ func fetchRollupRuleHistory(s *service, r *http.Request) (data interface{}, err
 	}
 	return models.NewRollupRuleSnapshots(hist), nil
 }
+
+func bulkUpdateRuleSet(s *service, r *http.Request) (data interface{}, err error) {
+	vars := mux.Vars(r)
+	var req bulkRequest
+	if err := parseRequest(&req, r.Body); err != nil {
+		return nil, err
+	}
+	uOpts, err := s.newUpdateOptions(r)
+	if err != nil {
+		return nil, err
+	}
+
+	originalRS, err := s.store.FetchRuleSet(vars[namespaceIDVar])
+	if err != nil {
+		return nil, err
+	}
+	if req.RuleSetVersion != originalRS.Version() {
+		return nil, NewConflictError(
+			fmt.Sprintf("can not update namespace %s, version mismatch", vars[namespaceIDVar]),
+		)
+	}
+
+	newRS, err := applyChangesToRuleSet(
+		req.RuleSetChanges,
+		originalRS.ToMutableRuleSet().Clone(),
+		uOpts,
+		s.updateHelper,
+	)
+	if err != nil {
+		return nil, NewBadInputError(err.Error())
+	}
+
+	rs, err := s.store.UpdateRuleSet(newRS)
+	if err != nil {
+		return nil, err
+	}
+	latest, err := rs.Latest()
+	if err != nil {
+		return nil, err
+	}
+	serializableRuleSet := models.NewRuleSet(latest)
+
+	return serializableRuleSet, nil
+}
+
+func applyChangesToRuleSet(
+	rsc changes.RuleSetChanges,
+	ruleSet rules.MutableRuleSet,
+	uOpts store.UpdateOptions,
+	helper rules.RuleSetUpdateHelper,
+) (rules.MutableRuleSet, error) {
+
+	if len(rsc.MappingRuleChanges) == 0 &&
```

Comment:
It might be worth adding a method to changes.RuleSetChanges that returns a bool that indicates whether there are any changes present. If the ruleset structure changes (and I think it may with the multi rollup rule stuff), it'd be good to have this logic contained in one place rather than updating different areas. It's also cleaner IMO.

Reply:
Potentially. I will have to think about this one because even if a ruleset structure changes its not 100% that the change object will change very much.